File: rcwa/scatter_matrix.py
import numpy as np
from numpy.linalg import inv
from rcwa.parameter import Parameter

class ScatterMatrix:

    def __init__(self):
        self.S11: np.ndarray = None
        self.S12: np.ndarray = None
        self.S21: np.ndarray = None
        self.S22: np.ndarray = None

    def clone_me(self) -> "ScatterMatrix":
        clone = ScatterMatrix()
        clone.S11 = np.copy(self.S11)
        clone.S12 = np.copy(self.S12)
        clone.S21 = np.copy(self.S21)
        clone.S22 = np.copy(self.S22)
        return clone

    @staticmethod
    def redheffer_star_product(SA: "ScatterMatrix", SB: "ScatterMatrix") -> "ScatterMatrix":
        dim = SA.S11.shape[0]
        I = np.eye(dim, dtype=Parameter.dtype)       
        SAB = ScatterMatrix()
        
        
        bracket_1 = inv(I - (SB.S11 @ SA.S22))
        bracket_2 = inv(I - (SA.S22 @ SB.S11))
        #SAB11        
        second = SA.S12 @ bracket_1 @ (SB.S11 @ SA.S21)
        SAB.S11 = SA.S11 + second

        #SAB12 
        SAB.S12 = SA.S12 @ bracket_1 @ SB.S12

        #SAB21
        SAB.S21 = SB.S21 @ bracket_2 @ SA.S21

        #SAB22
        second = SB.S21 @ bracket_2 @ (SA.S22 @ SB.S12)
        SAB.S22 = SB.S22 + second

        return SAB
    
    @staticmethod
    def unity(dim_Sij: int) -> "ScatterMatrix":
        S = ScatterMatrix()
        S.S11 = np.zeros((dim_Sij, dim_Sij), dtype=Parameter.dtype)
        S.S21 = np.eye(dim_Sij, dtype=Parameter.dtype)
        S.S12 = np.eye(dim_Sij, dtype=Parameter.dtype)
        S.S22 = np.zeros((dim_Sij, dim_Sij), dtype=Parameter.dtype)
        return S

File: VolumeHologramEvaluationCPP/src/ScatterMatrix.cpp
#include "ScatterMatrix.h"

namespace rcwa
{

 ScatterMatrix::ScatterMatrix(){}

 ScatterMatrix ScatterMatrix::Unity(int dimSij) {
     ScatterMatrix s;
     s.s11 = arma::cx_mat(dimSij, dimSij, arma::fill::zeros);
     s.s21 = arma::eye<arma::cx_mat>(dimSij, dimSij);
     s.s12 = arma::eye<arma::cx_mat>(dimSij, dimSij);
     s.s22 = arma::cx_mat(dimSij, dimSij, arma::fill::zeros);
     return s;
 }

 ScatterMatrix ScatterMatrix::redhefferStarProduct(const ScatterMatrix& sa, const ScatterMatrix& sb) {
     int dim = sa.s11.n_rows;
     arma::cx_mat i = arma::eye<arma::cx_mat>(dim, dim);
     ScatterMatrix sab;
     
     arma::cx_mat bracket_1 = arma::inv(i - (sb.s11 * sa.s22));
     arma::cx_mat bracket_2 = arma::inv(i - (sa.s22 * sb.s11));

     
     //SAB11
     arma::cx_mat second = sa.s12 * bracket_1 * (sb.s11 * sa.s21);
     sab.s11 = sa.s11 + second;

     //SAB12
     sab.s12 = sa.s12 * bracket_1 * sb.s12;

     //SAB21
     sab.s21 = sb.s21 * bracket_2 * sa.s21;

     //SAB22
     second = sb.s21 * bracket_2 * (sa.s22 * sb.s12);
     sab.s22 = sb.s22 + second;
     
     return sab;
 }

 void  ScatterMatrix::print(const std::string& msg) const {
     
     std::cout << msg <<  std::endl;
     
     (s11).raw_print("s11");
     (s21).raw_print("s21");
     (s12).raw_print("s12");
     (s22).raw_print("s22");
 }

}